### PR TITLE
Better error when comparing indexers

### PR DESCRIPTION
**Pull request by mcon**
Previously you'd get a `TargetParameterCountException` when encountering an indexer, as indexers are properties with parameters, and the existing code assumed that no properties took parameters.

Ultimately there's no good way to compare indexers, as you can't obtain the full collection to compare, so this PR results in a better error in those cases.

### CODE REVIEW

File: src/Shouldly.Tests/ShouldBeEquivalentTo/ObjectScenario.cs
Comment (jnm2) [inline src/Shouldly.Tests/ShouldBeEquivalentTo/ObjectScenario.cs:299]: Normally I'd rather think of this as not being a test failure but being a Shouldly API usage failure and throw `NotSupportedException`. What do you think?

Also just to nitpick, what do you think about `Comparing types that have indexers is not supported.`? I like having full sentences including end marks for user-facing messages coming out of a library, including XML docs and exception messages.

File: src/Shouldly/ShouldlyExtensionMethods/ShouldBe/ObjectGraphTestExtensions.cs
Comment (jnm2) [inline src/Shouldly/ShouldlyExtensionMethods/ShouldBe/ObjectGraphTestExtensions.cs:161]: I have a slight preference for keeping related things together, and I think of the GetIndexParameters check and this `throw` as a logical unit. If you don't feel strongly, can you invert the `if` and remove the `else` nesting?

**Review by jnm2 — COMMENTED** [review]
I'd approve but I had some questions that I'd like input on. I'm open to going multiple directions.

File: src/Shouldly.Tests/ShouldBeEquivalentTo/ObjectScenario.cs
Comment (jnm2) [inline src/Shouldly.Tests/ShouldBeEquivalentTo/ObjectScenario.cs:298]: Could you add `.Message.ShouldBe(...)` on the end of this statement? It helps present a full picture of the user experience in the test code, and on rare occasions it has saved me from accidentally testing the wrong failure mode when the method throws the same exception type for two different reasons.

File: src/Shouldly/ShouldlyExtensionMethods/ShouldBe/ObjectGraphTestExtensions.cs
Comment (jnm2) [inline src/Shouldly/ShouldlyExtensionMethods/ShouldBe/ObjectGraphTestExtensions.cs:154]: Nit: could you add a period to make this user-facing message a complete sentence?

**Review by jnm2 — CHANGES_REQUESTED** [review]

**Review by jnm2 — APPROVED** [review]
Awesome! Thank you so much for getting this done!

**Comment by jnm2** [comment]
Uh oh, can't merge without resolving conflicts. Are you able to either rebase to fix the conflicts or merge the upstream master branch, creating a merge commit to fix the merge conflicts?

**Comment by mcon** [comment]
> Can you actually reopen and push/force push rather than starting a new PR? Thanks.

Slightly confused, looks like this is merged into master now, but the PR doesn't reflect that...

**Comment by jnm2** [comment]
Well look at that, that's weird. I did press the green Merge button, then the UI didn't refresh. I refreshed the page and it said it couldn't merge until conflicts were fixed, which I wasn't expecting. Looks like a UI glitch.

**Comment by jnm2** [comment]
Sorry about that. I guess no further action is necessary. If it bothers you that the PR appears to have been closed instead of merged, I'm willing to try to figure out how to reverse all this and merge properly. If not, I'll just put `[GitHub bug, actually merged]` in the PR title so I don't get confused when looking back.

**Comment by mcon** [comment]
Yeah, feels like some sort of GH bug... I would bother looking into it unless it happens again, modifying the title to avoid confusion sounds like the way to go.

Thanks for the review!

**Comment by jnm2** [comment]
I submitted a GitHub support ticket in case this PR is useful for them in tracking down a bug.

**Comment by jnm2** [comment]
Response:

> Thanks for your patience while we worked on this!
> 
> Our engineers have investigated this issue and determined that the mobile aspect here appears to be a red herring.
> 
> The pull request got into a bad state from a partial merge action in the application which can occur, albeit infrequently, for various reasons. The team are working on better ways to manage this error going forward, but I'm afraid we are unable to manipulate the current state of the pull request.
> 
> If you do experience this issue again please do let us know.

**Pull request closed without merging**
